packages/core/src/fields/types/file/index.ts

```diff
@@ -1,128 +1,117 @@
   ): FieldTypeFunc<ListTypeInfo> =>
-  () => {
-    if ((config as any).isIndexed === 'unique') {
+  meta => {
+    const storage = meta.assets.getStorage(config.storage);
+
+    if (!storage) {
+      throw new Error(
+        `${meta.listKey}.${meta.fieldKey} has storage set to ${config.storage} but there is no storage config under that key`
+      );
+    }
+
+    if ('isIndexed' in config) {
       throw Error("isIndexed: 'unique' is not a supported option for field type file");
     }
 
     return fieldType({
       kind: 'multi',
       fields: {
         filesize: { kind: 'scalar', scalar: 'Int', mode: 'optional' },
-        mode: { kind: 'scalar', scalar: 'String', mode: 'optional' },
         filename: { kind: 'scalar', scalar: 'String', mode: 'optional' },
       },
     })({
       ...config,
+      hooks: storage.preserve
+        ? config.hooks
+        : {
+            ...config.hooks,
+            async beforeOperation(args) {
+              await config.hooks?.beforeOperation?.(args);
+              if (args.operation === 'update' || args.operation === 'delete') {
+                const filenameKey = `${meta.fieldKey}_filename`;
+                const filename = args.item[filenameKey];
+
+                // This will occur on an update where a file already existed but has been
+                // changed, or on a delete, where there is no longer an item
+                if (
+                  (args.operation === 'delete' ||
+                    typeof args.resolvedData[meta.fieldKey].filename === 'string' ||
+                    args.resolvedData[meta.fieldKey].filename === null) &&
+                  typeof filename === 'string'
+                ) {
+                  await args.context.files(config.storage).deleteAtSource(filename);
+                }
+              }
+            },
+          },
       input: {
-        create: { arg: graphql.arg({ type: FileFieldInput }), resolve: inputResolver },
-        update: { arg: graphql.arg({ type: FileFieldInput }), resolve: inputResolver },
+        create: {
+          arg: graphql.arg({ type: FileFieldInput }),
+          resolve: (data, context) => inputResolver(config.storage, data, context),
+        },
+        update: {
+          arg: graphql.arg({ type: FileFieldInput }),
+          resolve: (data, context) => inputResolver(config.storage, data, context),
+        },
       },
       output: graphql.field({
         type: FileFieldOutput,
-        resolve({ value: { filesize, filename, mode } }) {
-          if (
-            filesize === null ||
-            filename === null ||
-            mode === null ||
-            (mode !== 'local' && mode !== 'cloud')
-          ) {
+        resolve({ value: { filesize, filename } }) {
+          if (filesize === null || filename === null) {
             return null;
           }
-          return { mode, filename, filesize };
+          return { filename, filesize, storage: config.storage };
```

Comment:
@mitchellhamilton you had a reason to return storage here yeah?

Reply:
Needed for the URL resolver